appbuilder/services/npm-service.ts

```diff
@@ -0,0 +1,233 @@
+
+export class NpmService implements INpmService {
+	private static TYPES_DIRECTORY = "@types/";
+	private static TNS_CORE_MODULES_DEFINITION_FILE_NAME = `${constants.TNS_CORE_MODULES}${constants.FileExtensions.TYPESCRIPT_DEFINITION_FILE}`;
+	private static NPM_REGISTRY_URL = "https://registry.npmjs.org";
+
+	private _npmExecutableName: string;
+
+	constructor(private $childProcess: IChildProcess,
+		private $errors: IErrors,
+		private $fs: IFileSystem,
+		private $hostInfo: IHostInfo,
+		private $httpClient: Server.IHttpClient,
+		private $logger: ILogger,
+		private $projectConstants: Project.IConstants) { }
+
+	@exportedPromise("npmService")
+	public install(projectDir: string, dependencyToInstall?: INpmDependency): IFuture<INpmInstallResult> {
+		return (() => {
+			let npmInstallResult: INpmInstallResult = {};
+
+			if (dependencyToInstall) {
+				npmInstallResult.result = {
+					isInstalled: false,
+					isTypesInstalled: false
+				};;
+
+				try {
+					this.npmInstall(projectDir, dependencyToInstall.name, dependencyToInstall.version, ["--save", "--save-exact"]).wait();
+					npmInstallResult.result.isInstalled = true;
+				} catch (err) {
+					npmInstallResult.error = err;
+				}
+
+				if (dependencyToInstall.installTypes && npmInstallResult.result.isInstalled && this.hasTypesForDependency(dependencyToInstall.name).wait()) {
+					try {
+						this.installTypingsForDependency(projectDir, dependencyToInstall.name).wait();
+						npmInstallResult.result.isTypesInstalled = true;
+					} catch (err) {
+						npmInstallResult.error = err;
+					}
+				}
+			} else {
+				try {
+					this.npmPrune(projectDir).wait();
+					this.npmInstall(projectDir).wait();
+				} catch (err) {
+					npmInstallResult.error = err;
+				}
+			}
+
+			this.generateReferencesFile(projectDir).wait();
+
+			return npmInstallResult;
+		}).future<INpmInstallResult>()();
+	}
+
+	@exportedPromise("npmService")
+	public uninstall(projectDir: string, dependency: string): IFuture<void> {
+		return (() => {
+			let packageJsonContent = this.getPackageJsonContent(projectDir).wait();
+
+			if (packageJsonContent && packageJsonContent.dependencies && packageJsonContent.dependencies[dependency]) {
+				this.npmUninstall(projectDir, dependency, ["--save"]).wait();
+			}
+
+			if (packageJsonContent && packageJsonContent.devDependencies && packageJsonContent.devDependencies[`${NpmService.TYPES_DIRECTORY}${dependency}`]) {
+				this.npmUninstall(projectDir, `${NpmService.TYPES_DIRECTORY}${dependency}`, ["--save-dev"]).wait();
+			}
+		}).future<void>()();
+	}
+
+	public getPackageJsonFromNpmRegistry(packageName: string, version?: string): IFuture<any> {
+		return (() => {
+			let packageJsonContent: any;
+			version = version || "latest";
+			try {
+				let url = this.buildNpmRegistryUrl(packageName, version);
+				// This call will return error with message '{}' in case there's no such package.
+				let result = this.$httpClient.httpRequest(url).wait().body;
```

Comment:
Our httpRequest method will throw error when the status code is 404 and will set the message to the content of the response. The content of the response is empty object. So yes, the method will never return empty object, but it will throw an error with message ={} and we'll go to the `catch`